dask_planner/src/sql/schema.rs

```diff
@@ -29,7 +32,24 @@ impl DaskSchema {
         self.tables.insert(table.name.clone(), table);
     }
 
-    pub fn add_function(&mut self, function: function::DaskFunction) {
-        self.functions.insert(function.name.clone(), function);
+    pub fn add_or_overload_function(
+        &mut self,
+        name: String,
+        input_types: Vec<PyDataType>,
+        return_type: PyDataType,
+    ) {
+        self.functions
+            .entry(name.clone())
+            .and_modify(|e| {
+                (*e).lock()
+                    .unwrap()
+                    .return_types
+                    .insert(input_types.clone(), return_type.clone());
+            })
+            .or_insert(Arc::new(Mutex::new(DaskFunction::new(
+                name,
+                input_types,
+                return_type,
+            ))));
```

Comment:
We can use `or_insert_with` with a closure here so that we only create a new `DaskFunction` if we don't find one in the map. Kudos to Clippy for this suggestion.

```suggestion
            .or_insert_with(|| {
                Arc::new(Mutex::new(DaskFunction::new(
                    name,
                    input_types,
                    return_type,
                )))
            });
```